### DISCUSSION TITLE
如何动态修改Field.component?

### DISCUSSION THREAD

**Discussion by xmsz** [alibaba/formily#3734] [Q&A 2.x] [answered]
背景
```
<Field
            name='targetReceivers.ids'
            decorator={[FormItem]}
             component={[
               targetType === 'rolePlan' ? RolePlanSelect : RoomSelect,
             ]}
          />
```
但是不管targetType怎么变化，页面上始终是RoomSelect这个组件

我通过
```
 form.current.setFieldState('targetReceivers.ids', (state) => {
            state.component = [
              targetType === 'rolePlan' ? <div>asd</div> : RoomSelect,
            ];
          });
``
也不行


我用之前的

```
{targetType === 'room' &&  <Field
            name='targetReceivers.ids'
            decorator={[FormItem]}
             component={[
              RoomSelect
             ]}
          />}
{targetType === 'rolePlan' &&  <Field
            name='targetReceivers.ids'
            decorator={[FormItem]}
             component={[
               RolePlanSelect
             ]}
          />}
```
也不行




我的需求就是不同的targetType，渲染不同的组件，要如何处理

**Comment by xiaosi-ss** [comment, marked as answer]
<img width="1499" alt="image" src="https://user-images.githubusercontent.com/7309546/220323752-d1957c0e-e550-4ff4-8a28-966b80f4ae21.png">
_想set啥 就set啥_

**Reply by xmsz** [reply]
可用 就是有点麻烦 要是可以
```
{targetType === 'room' && <Field
name='targetReceivers.ids'
decorator={[FormItem]}
component={[
RoomSelect
]}
/>}
{targetType === 'rolePlan' && <Field
name='targetReceivers.ids'
decorator={[FormItem]}
component={[
RolePlanSelect
]}
/>}
``` 
切换就好了


还有我想问一下，那我要动态切换validator呢？
因为RoomSelect和RolePlanSelect的validator是不同的

**Reply by xiaosi-ss** [reply]
![image](https://user-images.githubusercontent.com/7309546/220510761-775ac83f-2358-44c3-bed7-c2e1ecdb30e0.png)

**Comment by coolbob1998** [comment]
Field 组件不是受控的
![image](https://user-images.githubusercontent.com/39683861/220604320-f44150b6-e3a7-4db0-b759-01bf6d4dbc73.png)

**Reply by spt110** [reply]
响应式，Field组件是受控的
<img width="939" alt="image" src="https://user-images.githubusercontent.com/5217855/227695697-a541d967-28a1-4ba7-bead-5247e04e41d7.png">
